File: 0729-my-calendar-i/0729-my-calendar-i.py
import bisect

class MyCalendar:

    def __init__(self):
        self.starts = []
        self.ends = []
        

    def book(self, start: int, end: int) -> bool:

        if start in self.starts:
            return False
        
        if not self.starts:
            bisect.insort(self.starts, start)
            bisect.insort(self.ends, end)
            return True

        if start >= self.ends[-1] or end <= self.starts[0]:
            bisect.insort(self.starts, start)
            bisect.insort(self.ends, end)
            return True
        pos1 = bisect.bisect_right(self.ends, start)
        pos2 = bisect.bisect_left(self.starts, end)
        if pos1 == pos2:
            bisect.insort(self.starts, start)
            bisect.insort(self.ends, end)
            return True
        return False

        


# Your MyCalendar object will be instantiated and called as such:
# obj = MyCalendar()
# param_1 = obj.book(start,end)

File: 0729-my-calendar-i/0729-my-calendar-i.cpp
class MyCalendar {
public:
    set<pair<int, int>> bookings; // Store pairs of (start, end) times

    MyCalendar() {
    }

    bool book(int start, int end) {
        // Check if the new booking overlaps with existing bookings
        for (const auto& booking : bookings) {
            // If the new booking overlaps with an existing booking, return false
            if (start < booking.second && end > booking.first) {
                return false; // Overlapping condition
            }
        }
        
        // If no overlap, insert the new booking
        bookings.insert({start, end});
        return true;
    }
};
/**
 * Your MyCalendar object will be instantiated and called as such:
 * MyCalendar* obj = new MyCalendar();
 * bool param_1 = obj->book(start,end);
 */
